Extend pipeline to combine results in one tsv file

krish8484 opened the pull request:
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

All the motevo results are encapsulated in one tsv file

Fixes #15 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not reduced the existing code coverage

AngryMaciek commented on tests/integration/expected_output.md5:
```diff
@@ -20,3 +20,5 @@ fe56b8577bada06fea45bb8835e7d7ee ./tests/integration/output/motevo/motif_HNRNPF_
 faa2f9e3e8f432575e281878d93a85a6 ./tests/integration/output/motevo/motif_HNRNPF_824/posteriors
 c0a4a9155ec7958b00bd3140e97c6dc9 ./tests/integration/output/motevo/motif_HNRNPF_823/posterior_sites
 20e7921795665bbe24e585e741705634 ./tests/integration/output/motevo/motif_HNRNPF_823/posteriors
+bb5480584d2aba7c8cfae0d5629deba9 ./tests/integration/output/combined_MotEvo_results.tsv
+0f62e3b0dc5bc8b048a021632ace92fe ./tests/integration/output/combined_MotEvo_results.csv
```
There is also a corresponding `expected_output.txt` with the list of files, you should update that one as well.

AngryMaciek commented on workflow/Snakefile:
```diff
@@ -37,16 +37,13 @@ rule all:
     Target rule with final output of the pipeline
     """
     input:
-        PWM_input_dir = config["pwm_directory"],
-        DIR_MotEvo_results_pwm = expand(
-            os.path.join(
-                "{outdir}",
-                "motevo",
-                "{pwm_dir}"
-            ),
-            outdir = config["outdir"],
-            pwm_dir = gather_motifs_names(config["pwm_directory"])
```
As this is the input to your rule we do not need it in the last rule of the workflow anymore.

AngryMaciek commented on workflow/Snakefile:
```diff
@@ -282,3 +279,24 @@ rule run_MotEvo_analysis:
+        ),
+        outdir = config["outdir"]
+        ),
+    params: "posterior_sites"
+    script: "parse.py"
```
Now I realised that if we call it like this then the script depends on the content of the `snakemake.`. This is bad, as we would not be able to call it outside of the pipeline ever (implying no unit tests as well).
Please change it to the `bash` call as for the other rules.

AngryMaciek replied:
You will have to pass arguments to the script explicitly, please take a look at the file `format-ATtRACT-motifs.py` to see how to catch CLI arguments from Python.

AngryMaciek commented on workflow/parse.py:
```diff
@@ -0,0 +1,64 @@
+tabb['binding_region'] = []
+tabb['binding_sequence'] = []
+tabb['binding_energy'] = []
+tabb['fast_record'] =[] 
```
the format is called `FASTA`, it would be better to add the `a` and we have the full name here.

AngryMaciek commented on workflow/parse.py:
```diff
@@ -0,0 +1,64 @@
+import os 
```
Please add your authorship and License (Apache 2.0) on top, you may look it up in one of my other scripts to notice how I do it.

AngryMaciek commented on workflow/parse.py:
```diff
@@ -0,0 +1,64 @@
+tabb['binding_energy'] = []
+tabb['fast_record'] =[] 
+
+print(snakemake.input)
+print(snakemake.params)
```
There will not be here, as in the previous comment - we have to make a script independent of the workflow.

AngryMaciek commented on workflow/parse.py:
```diff
@@ -0,0 +1,64 @@
+
+dirrs = snakemake.input
+outdir = str(snakemake.output)
+print(outdir)
```
Please remove all such `print` checks in the final version. We want the code clean.

AngryMaciek commented on workflow/parse.py:
```diff
@@ -0,0 +1,64 @@
+
+    i = 0
+    j = 0
+    file = open(filename, "r")
```
Your code is not commented at all. Granted, these are trivial things but here you could inject 2,3 lines of comment to say how do you iterate over what et cetera.

AngryMaciek commented on workflow/Snakefile:
```diff
@@ -282,3 +279,24 @@ rule run_MotEvo_analysis:
         """
+
+rule combine_MotEvo_results:
+    input: 
```
You are missing a rule description, check out the other rules in the workflow.

AngryMaciek replied:
This is OK now,

AngryMaciek commented:
@krish8484 
Apart from what is in the code comments, I have unchecked some boxes from the checklist that still need to be addressed. As you change how the script is called you will also have to add unit tests for your script (the same way I have for another one). In order to generate dummy input for the unit test just run the pipeline once, copy the main directory with MotEvo results (per motif directories) and remove all of them except, let's say, arbitrarily chosen 10 (so that the test is rather small).

krish8484 commented:
Okay, i had done it the wrong way, will fix it thanks!

krish8484 commented:
All the changed mentioned above have been implemented. Both the `md5sum --check` commands are returning `OK` on my local machine, but failing on the travis server. I am not understanding the reason.

AngryMaciek commented on .travis.yml:
```diff
@@ -48,3 +48,10 @@ script:
   - bash tests/integration/execution/snakemake_local_run_conda_environments.sh
   - md5sum --check tests/integration/expected_output.md5
+  # Test parse.py script
+  - >
+    python scripts/parse.py
+    --indir tests/unit/combine_results/motif_HNRNPF_820 tests/unit/combine_results/motif_HNRNPF_821 tests/unit/combine_results/motif_HNRNPF_822 tests/unit/combine_results/motif_HNRNPF_823 tests/unit/combine_results/motif_HNRNPF_824
+    --filename posterior_sites
+    --outfile tests/unit/combine_results/combined_MotEvo_results.tsv
+  - md5sum --check tests/unit/combine_results/expected_output.md5 
```
This is OK, but please put it before the integration test. As a good practice: All unit tests should be executed and passed before the integration test starts.

AngryMaciek commented on envs/main.yml:
```diff
@@ -28,5 +28,4 @@
     - pyyaml=3.12
     - snakemake=5.19.0
     - unzip=6.0
-
```
If the other YAML files have space here then please leave it as it was, for consistency.

AngryMaciek commented on scripts/parse.py:
```diff
@@ -0,0 +1,112 @@
+##############################################################################
+#   AUTHOR: Krish Agarwal
+#   AFFILIATION: University_of_Basel
+#   AFFILIATION: Swiss_Institute_of_Bioinformatics
```
As you are not officially an SIB member I would remove this line but keep only University of Basel, as you are working in our research group now.

AngryMaciek commented on tests/integration/output/local_log/prepare_MotEvo_parameters.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
This is a log file with execution time - it should not be under version control. Miliseconds of execution might change between pipeline runs but that is OK.

AngryMaciek replied:
The same applies to two streams below: empty `stderr` and `stdout`

AngryMaciek commented on tests/integration/output/local_log/prepare_sequence_for_MotEvo.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
The same situation here - this is a log file

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_820.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
The same situation here - this is a log file

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_821.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
The same situation here - this is a log file

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_822.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
The same situation here - this is a log file.
This applies to all the `stderr` and `stdout` files as well(!)

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_823.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
The same situation here - this is a log file

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_824.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
The same situation here - this is a log file

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_825.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
The same situation here - this is a log file

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_826.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
+0.0984	0:00:00	3.59	10.85	0.18	0.78	0.00	0.00	0.00
```
The same situation here - this is a log file

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_827.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
The same situation here - this is a log file

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_828.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
The same situation here - this is a log file

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_829.benchmark.log:
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
The same situation here - this is a log file

AngryMaciek commented:
I have made some minor comments, please remove unnecessary files first, before I git into the script itself.
Also, please remove all `.snakemake_timestamp ` files. There are internal time stamps that should not be under version control;

AngryMaciek commented on tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_820.stdout.log:
```diff
@@ -0,0 +1,53 @@
+letter A val is 0.250000
```
Please remove ALL `stdout` and `stderr` files from the repository. These are logs and should not be kept under version control.  
It makes me wonder - why did you add these exact ones in the first place? These streams are related to `run_MotEvo_analysis` rule, which I already covered...

AngryMaciek commented on tests/integration/output/sequence.fasta:
```diff
@@ -0,0 +1,2 @@
+>>MOTEVO_input-sequence
+ATGTGAGTGAAGTGTGGGAAAGATGACTCGATATATCTGGATGCTAGGGATCGGATGGCGATACG
```
Why do you need this file under version control of the integration test?

AngryMaciek commented on workflow/Snakefile:
```diff
@@ -282,3 +279,29 @@ rule run_MotEvo_analysis:
+    params:
+        Input_filename = "posterior_sites"
+    shell: 
+        "python ../../../scripts/parse.py --indir {input.DIR_MotEvo_results_pwm}  --filename {params.Input_filename} --outfile {output.combined_MotEvo_results}"
```
Please break the line with `\` characterl (as in the other rule); there is a line width limit

krish8484 commented:
The log files were auto generated after running the pipeline, you are right they were not needed, so I deleted them now.
The tests are passing sometime and failing other time because the order of input directories that the snakemake file receives is not the same all the time, so the order of tsv file changes according to the input.
I have implemented the other changes.

AngryMaciek commented:
> The integration tests are passing sometime and failing other time because the order of input directories that the snakemake file receives is not the same all the time, so the order of tsv file changes according to the input. I think unit tests do a good job of checking the script.

OK, thanks! Looks good now. But before I merge we need to take care about this randomness effect. Reproducibility is a very important aspect of data analysis and we cannot leave any non-deterministic effects like that. That is a easy fix, though - just sort the list of paths to the files with `sorted()` buitl-in function. Then the order will always be the same.

AngryMaciek commented on .travis.yml:
```diff
@@ -42,9 +42,17 @@ script:
+  # Test parse.py script
+  - >
+    python scripts/parse.py
+    --indir tests/unit/combine_results/motif_HNRNPF_820 tests/unit/combine_results/motif_HNRNPF_821 tests/unit/combine_results/motif_HNRNPF_822 tests/unit/combine_results/motif_HNRNPF_823 tests/unit/combine_results/motif_HNRNPF_824
```
could you please change the argument to `input-directories` explicite? :)

krish8484 replied:
Done!

AngryMaciek commented on scripts/parse.py:
```diff
@@ -0,0 +1,111 @@
+        i = i + 1
+
+df = pd.DataFrame({key: pd.Series(value) for key, value in tabb.items()})
+df.to_csv(rel_outfile_csv, index=False)
+
+
+with open(rel_outfile_csv, 'r') as csvin, open(rel_outfile, 'w') as tsvout:
+    csvin = csv.reader(csvin)
+    tsvout = csv.writer(tsvout, delimiter='\t')
+
+    for row in csvin:
+        tsvout.writerow(row) #printing row by row in tsv
```
We do not need two output files, just the `tsv` is enough.
You may use: `df.to_csv(rel_outfile_csv, index=False, sep="\t)`

AngryMaciek replied:
This applies to all the CI - you should check just the `tsv file`

krish8484 replied:
Done!

AngryMaciek commented on scripts/parse.py:
```diff
@@ -0,0 +1,111 @@
+
+    filename = os.path.relpath(abs_path)
+
+    i = 0
+    j = 0
+    file = open(filename, "r")
+    for each in file:
+        each_word = each.split(' ')
```
You still do not have any comments on your code...

krish8484 replied:
Added!

AngryMaciek commented on tests/unit/combine_results/combined_MotEvo_results.csv:
```diff
@@ -0,0 +1,22 @@
+strand,binding_position,binding_posterior,pwm_id,binding_region,binding_sequence,binding_energy,fasta_record
++,14-18,0.015291,HNRNPF_820,"MOTEVO_input-sequence
+",GTGGG,2.51049,"MOTEVO_input-sequence
```
I am very confused by the information here... why is every record separated into two rows? Each record should be one line (i.e. one row) in the result table. The records should match the header of the table.

AngryMaciek commented on tests/unit/combine_results/combined_MotEvo_results.tsv:
```diff
@@ -0,0 +1,22 @@
+strand	binding_position	binding_posterior	pwm_id	binding_region	binding_sequence	binding_energy	fasta_record
++	14-18	0.015291	HNRNPF_820	"MOTEVO_input-sequence
+"	GTGGG	2.51049	"MOTEVO_input-sequence
+"
++	6-10	0.017773	HNRNPF_823	"MOTEVO_input-sequence
+"	AGTGA	2.51049	"MOTEVO_input-sequence
+"
++	15-19	0.018685	HNRNPF_823	"MOTEVO_input-sequence
+"	TGGGA	2.51049	"MOTEVO_input-sequence
+"
++	46-50	0.674978	HNRNPF_823	"MOTEVO_input-sequence
+"	AGGGA	6.71171	"MOTEVO_input-sequence
+"
++	15-19	0.560144	HNRNPF_824	"MOTEVO_input-sequence
+"	TGGGA	6.71171	"MOTEVO_input-sequence
+"
++	46-50	0.030141	HNRNPF_824	"MOTEVO_input-sequence
+"	AGGGA	2.51049	"MOTEVO_input-sequence
+"
++	51-55	0.014195	HNRNPF_824	"MOTEVO_input-sequence
+"	TCGGA	2.51049	"MOTEVO_input-sequence
```
I am very confused by the information here... why is every record separated into two rows? Each record should be one line (i.e. one row) in the result table. The records should match the header of the table.

krish8484 replied:
That is because you are viewing the tsv file in GitHub. Have you checked it using LibreOffice?

AngryMaciek replied:
Oh, OK, thanks! GitHub confused me...

AngryMaciek commented on tests/unit/combine_results/combined_MotEvo_results.tsv:
```diff
@@ -0,0 +1,22 @@
+strand	binding_position	binding_posterior	pwm_id	binding_region	binding_sequence	binding_energy	fasta_record
```
I would rearrange the order of columns here a little:
`pwm_id binding_position, binding_sequence	binding_posterior	binding_energy`

We do not need `strand` or `fasta_record` for now. Please keep them in the code but just do not use them to generate the table. The same applies to `binding_region` - note that it is the same as `fasta_record`

krish8484 replied:
Done!

AngryMaciek commented:
OK, looks very good now. ╰(▔∀▔)╯  
  
As I was going through the files in detail I also noticed that you do not have a *conda* virtual environment for your rule, right? Each snakemake rule is executed in a separated, dedicated virtual environment which contains all the packages needed. They are automacitally built by snakemake during the pipeline execution, built based on the YAML files. Take a look at other rules in the workflow. Could you please add a YAML file with specifications related to your script and a proper field in the Snakefile?

AngryMaciek commented on workflow/Snakefile:
```diff
@@ -282,3 +279,32 @@ rule run_MotEvo_analysis:
+    params:
+        Input_filename = "posterior_sites"
+    shell: 
+        "python ../../../scripts/parse.py \
```
This is a comment for @AngryMaciek :
Double check how to encode relative paths, also - check the execution bash scripts for `cwd` there

AngryMaciek replied:
This is a more complex issue - I will have to create a separate feature branch for this fix.

AngryMaciek commented:
Please also run a `rulegraph run` test to generate new image of the workflow that will be displayed in the README (it should contain your rule). The new image file should substitute the old one in `images` directory.

krish8484 commented on .travis.yml:
```diff
@@ -42,9 +42,17 @@ script:
   - md5sum --check tests/unit/format-ATtRACT-motifs/expected_output_mmu.md5
+  # Test parse.py script
+  - >
+    python scripts/parse.py
```
done, i have changed the path

krish8484 commented:
Updated the image of the workflow, added env file, formatted using `black` and changed the path of `parse.py`
